x/twap/listeners_test.go

```diff
@@ -3,35 +3,93 @@ package twap_test
+// TestAfterPoolCreatedHook tests if internal tracking logic has been triggered correctly,
+// and the corerct state entries have been created upon pool creation.
+// This test includes test cases for swapping on the same block with pool creation.
+func (s *TestSuite) TestAfterPoolCreatedHook() {
+	two_asset_coin := defaultUniV2Coins
+	three_asset_coins := defaultThreeAssetCoins
+
+	tests := []struct {
+		name      string
+		poolCoins sdk.Coins
+		// if this field is set true, we swap in the same block with pool creation
+		runSwap bool
+	}{
+		{
+			"Uni2 Pool, no swap on pool creation block",
+			two_asset_coin,
+			false,
+		},
+		{
+			"Uni2 Pool, swap on pool creation block",
+			two_asset_coin,
+			true,
+		},
+		{
+			"Three asset balancer pool, no swap on pool creation block",
+			three_asset_coins,
+			false,
+		},
+		{
+			"Three asset balancer pool, no swap on pool creation block",
+			three_asset_coins,
+			true,
+		},
+	}
+
+	for _, tc := range tests {
+		s.SetupTest()
+		s.Run(tc.name, func() {
+			var poolId uint64
+			// prepare pool according to test case
+			poolId = s.PrepareBalancerPoolWithCoins(tc.poolCoins...)
+
+			if tc.runSwap {
+				s.RunBasicSwap(poolId)
+			}
+
+			timeBeforeBeginBlock := s.Ctx.BlockTime()
+			s.twapkeeper.EndBlock(s.Ctx)
+			s.BeginNewBlock(false)
+
+			// check that creating a pool saved a new state of changed pools
+			s.Require().Equal([]uint64{poolId}, s.twapkeeper.GetChangedPools(s.Ctx))
+
+			// check that all twap records have been created for all denom pairs in pool
+			twapRecords, err := s.twapkeeper.GetAllMostRecentRecordsForPool(s.Ctx, poolId)
+			s.Require().NoError(err)
+			denoms, err := s.App.GAMMKeeper.GetPoolDenoms(s.Ctx, poolId)
+			s.Require().NoError(err)
+			denomPairs0, _ := types.GetAllUniqueDenomPairs(denoms)
+			s.Require().Equal(len(denomPairs0), len(twapRecords))
+
+			// check on the individual twap records
+			for _, twapRecord := range twapRecords {
+				s.Require().Equal(poolId, twapRecord.PoolId)
+				s.Require().Equal(sdk.ZeroDec(), twapRecord.P0ArithmeticTwapAccumulator)
+				s.Require().Equal(sdk.ZeroDec(), twapRecord.P1ArithmeticTwapAccumulator)
+
+				asset0sp, err := s.App.GAMMKeeper.CalculateSpotPrice(s.Ctx, poolId, twapRecord.Asset0Denom, twapRecord.Asset1Denom)
+				s.Require().NoError(err)
+				s.Require().Equal(asset0sp, twapRecord.P0LastSpotPrice)
+				s.Require().Equal(timeBeforeBeginBlock, twapRecord.Time)
+
+				// check that we have same state entry
+				storeRepresentationRecord, err := s.twapkeeper.GetMostRecentRecordStoreRepresentation(s.Ctx, poolId, twapRecord.Asset0Denom, twapRecord.Asset1Denom)
+				s.Require().Equal(twapRecord, storeRepresentationRecord)
+
+				twapRecordBeforeTime, err := s.twapkeeper.GetRecordAtOrBeforeTime(s.Ctx, poolId, s.Ctx.BlockTime(), twapRecord.Asset0Denom, twapRecord.Asset1Denom)
+				s.Require().NoError(err)
+				s.Require().Equal(twapRecord, twapRecordBeforeTime)
```

Comment:
Can we just construct the expected TWAP record, and then check equality with one representation?

I don't think we need to check all representations, or if we do that should be a separate method - its a guarantee elsewhere in the code that all methods update together.